### PR TITLE
Create dynamic configuration phase

### PR DESCRIPTION
**Pull request by Rezney**
This commit adds a new "configuration" field into a workflow.
When a configuration model is added to the workflow and there is
an actor which produces it, the producing actor is run first as
part of the dynamically added phase and produces the config model
for consumption for all other actors, together with implementing
self.configuration api in every actor.

Example:  

```python
class FoobarWorkflow(Workflow):
    name = 'FoobarWorkflow'
    tag =  FoobarWorkflowTag
    short_name = 'foobar_workflow'
    description = '''No description has been provided ...'''
    configuration = FoobarConfig

    class FirstPhaseName(Phase):
       name = 'first_phase_name'
       filter = TagFilter(FirstPhaseTag)
       policies = Policies(Policies.Errors.FailPhase,
                           Policies.Retry.Phase)
       flags = Flags()
```

Then produce the configuration model using a "non-phase"
actor (see tags field):

```python
class FoobarConfigProducer(Actor):
"""
No documentation has been provided ... 
"""

    name = 'foobar_config_producer'
    consumes = ()
    produces = (FoobarConfig,)
    tags = (FoobarWorkflowTag,)

    def process(self):
        self.produce(FoobarConfig())
```

kudos to @vinzenz

Having the configuration will be beneficial e.g. when working on minor versioning.

### CODE REVIEW

File: leapp/workflows/__init__.py [hunk 1]
```diff
@@ -39,6 +39,12 @@ def __new__(mcs, name, bases, attrs):
         return klass
 
 
+
```
Comment (pirat89) [inline leapp/workflows/__init__.py]:
```suggestion
```

File: leapp/workflows/__init__.py [hunk 2]
```diff
@@ -39,6 +39,12 @@ def __new__(mcs, name, bases, attrs):
...
+class _ConfigPhase(Phase):
+    name = 'configuration_phase'
+
+
+
```
Comment (pirat89) [inline leapp/workflows/__init__.py]:
```suggestion
```

**Comment by pirat89** [comment]
Rebase please. A lot of important commits are missing here to be able to test it with other content.

**Comment by Rezney** [comment]
@pirat89 thanks for checking this and sorry for delay. Rebased, commit message fixed, blank spaces gone...

**Comment by Rezney** [comment]
Started hitting this when testing the PR:

```python
May 17 06:46:10 localhost upgrade[387]: Process Process-57:
May 17 06:46:10 localhost upgrade[387]: Traceback (most recent call last):
May 17 06:46:10 localhost upgrade[387]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
May 17 06:46:10 localhost upgrade[387]:     self.run()
May 17 06:46:10 localhost upgrade[387]:   File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
May 17 06:46:10 localhost upgrade[387]:     self._target(*self._args, **self._kwargs)
May 17 06:46:10 localhost upgrade[387]:   File "/root/tmp_leapp_py3/leapp/repository/actor_definition.py", line 26, in inspect_actor
May 17 06:46:10 localhost upgrade[387]:     definition.load()
May 17 06:46:10 localhost upgrade[387]:   File "/root/tmp_leapp_py3/leapp/repository/actor_definition.py", line 143, in load
May 17 06:46:10 localhost upgrade[387]:     self._module = importer.find_module(name).load_module(name)
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap_external>", line 399, in _check_name_wrapper
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap_external>", line 823, in load_module
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap_external>", line 682, in load_module
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap>", line 265, in _load_module_shim
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap>", line 684, in _load
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap_external>", line 678, in exec_module
May 17 06:46:10 localhost upgrade[387]:   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
May 17 06:46:10 localhost upgrade[387]:   File "/etc/leapp/repos.d/system_upgrade/el7toel8/actors/rpmscanner/actor.py", line 1, in <module>
May 17 06:46:10 localhost upgrade[387]:     import yum
May 17 06:46:10 localhost upgrade[387]: ModuleNotFoundError: No module named 'yum'
May 17 06:46:10 localhost upgrade[387]: 2019-05-17 06:46:10.319 ERROR    PID: 3 leapp.repository.system_upgrade_el7toel8: Process inspecting actor in actors/rpmscanner failed with 1
May 17 06:46:10 localhost upgrade[387]: Inspection of actor in actors/rpmscanner failed
```

but found out it is broken master :)

**Comment by vojtechsokol** [comment]
leapp-ci build

**Comment by pirat89** [comment]
@Rezney it should be fixed now. I will review the PR later this day

**Comment by Rezney** [comment]
@pirat89 rebased...

**Comment by pirat89** [comment]
> @Rezney it should be fixed now. I will review the PR later this day

:skull:  ... but I am finally working on it :crossed_fingers:

File: leapp/workflows/__init__.py [hunk 3]
```diff
@@ -77,12 +84,25 @@ def __init__(self, logger=None, auto_reboot=False):
...
 
+        if self.configuration:
+            config_actors = [actor for actor in self.tag.actors if self.configuration in actor.produces]
+            if config_actors and len(config_actors) == 1:
+                self._phase_actors.append((
+                    _ConfigPhase,
+                    PhaseActors((), 'Before'),
+                    PhaseActors(tuple(config_actors), 'Main'),
+                    PhaseActors((), 'After')))
+
```
Comment (pirat89) [inline leapp/workflows/__init__.py]: Missing the `elif` statement. It should fail when there are more actors producing the message.

**Comment by pirat89** [comment]
Additionally, still it is discussed how to resolve the thing for snactor. With the current implementation, the actor processed just by `snactor run` will not behave as it is expected.

File: leapp/workflows/__init__.py (hunk 3 above)
Comment (Rezney) [reply leapp/workflows/__init__.py]: Done

**Comment by Rezney** [comment]
Rebased. @pirat89 @vinzenz let's discuss what to do with snactor...

File: leapp/actors/__init__.py [hunk 4]
```diff
@@ -113,6 +122,13 @@ def show_message(self, message):
         """
         self._messaging.show_message(message)
 
+    @property
+    def get_config(self):
+        """ Returns the config model generated by specific workflow configuration actor"""
```
Comment (pirat89) [inline leapp/actors/__init__.py]:
```suggestion
        """
        Returns the config model generated by specific workflow configuration actor.
        """
```

File: leapp/actors/__init__.py [hunk 5]
```diff
@@ -84,13 +84,22 @@ def serialize(self):
...
         }
 
-    def __init__(self, messaging=None, logger=None):
+    def __init__(self, messaging=None, logger=None, config_model=None):
+        self.configuration = None
+        """
+        Instance a workflow defined configuration model if available. This depends on the definition of such a
+        configuration model being defined by the workflow and an actor that provides such a message.
+        """
+
```
Comment (pirat89) [inline leapp/actors/__init__.py]:
```
        """
        Instance a workflow defined configuration model if available.

        This depends on the definition of such a configuration model being defined by the workflow
        and an actor that provides such a message.
        """
```

File: leapp/actors/__init__.py [hunk 6]
```diff
@@ -84,13 +84,22 @@ def serialize(self):
             'dialogs': [d.serialize() for d in self.dialogs],
         }
 
-    def __init__(self, messaging=None, logger=None):
+    def __init__(self, messaging=None, logger=None, config_model=None):
+        self.configuration = None
```
Comment (pirat89) [inline leapp/actors/__init__.py]: privateee

```suggestion
        self._configuration = None
```

**Comment by Rezney** [comment]
@pirat89 your comments should be resolved. Let me know if anything else is needed...

File: leapp/actors/__init__.py [hunk 7]
```diff
@@ -113,6 +123,13 @@ def show_message(self, message):
         """
         self._messaging.show_message(message)
 
+    @property
+    def get_config(self):
```
Comment (vinzenz) [inline leapp/actors/__init__.py]:
```suggestion
    def configuration(self):
```

File: leapp/exceptions.py [hunk 8]
```diff
@@ -63,6 +63,13 @@ def __init__(self, path):
             'Multiple actors found in {path}. Inspection failed'.format(path=path))
 
 
+class MultipleConfigActorsError(LeappError):
```
Comment (vinzenz) [inline leapp/exceptions.py]: Where is `WorkflowConfigNotAvailable`?

File: leapp/actors/__init__.py (hunk 4 above)
Comment (pirat89) [reply leapp/actors/__init__.py]: @Rezney this one

File: leapp/exceptions.py (hunk 8 above)
Comment (Rezney) [reply leapp/exceptions.py]: Oops, not saved vscode file. Will fix that first thing tomorrow...

**Pull request merged**